File: examples/scala/src/main/scala/com/xebia/functional/xef/scala/auto/TopAttraction.scala
package com.xebia.functional.xef.scala.conversation

import com.xebia.functional.xef.scala.conversation.*
import io.circe.Decoder
import com.xebia.functional.xef.prompt.Prompt

private final case class City(name: String, country: String) derives SerialDescriptor, Decoder

private final case class TopAttractionWeather(city: City, temperature: Double, description: String) derives SerialDescriptor, Decoder

private final case class TopAttraction(city: City, attractionName: String, description: String, weather: TopAttractionWeather)
    derives SerialDescriptor,
      Decoder

@main def runTopAttraction: Unit =
  conversation {
    val nearbyTopAttraction: TopAttraction = prompt(Prompt("Top attraction in Cádiz, Spain."))
    println(
      s"""
        |The top attraction in ${nearbyTopAttraction.city.name} is ${nearbyTopAttraction.attractionName}.
        |Here's a brief description: ${nearbyTopAttraction.description}.
        |The weather in ${nearbyTopAttraction.city.name} is ${nearbyTopAttraction.weather.temperature} degrees Celsius and ${nearbyTopAttraction.weather.description}.
        |""".stripMargin
    )
  }

File: examples/kotlin/src/main/kotlin/com/xebia/functional/xef/auto/TopAttraction.kt
package com.xebia.functional.xef.auto

import com.xebia.functional.xef.auto.llm.openai.getOrElse
import com.xebia.functional.xef.auto.llm.openai.prompt
import kotlinx.serialization.Serializable

@Serializable
data class TopAttraction(val city: City, val attractionName: String, val description: String, val weather: Weather)

@Serializable
data class City(val name: String, val country: String)

@Serializable
data class Weather(val city: City, val temperature: Double, val description: String)

suspend fun main() = ai {
  val nearbyTopAttraction: TopAttraction = prompt("Top attraction in Cádiz, Spain.")
  println(
      """
          |The top attraction in ${nearbyTopAttraction.city.name} is ${nearbyTopAttraction.attractionName}. 
          |Here's a brief description: ${nearbyTopAttraction.description}.
          |The weather in ${nearbyTopAttraction.city.name} is ${nearbyTopAttraction.weather.temperature} degrees Celsius and ${nearbyTopAttraction.weather.description}.
          |""".trimMargin()
  )
}.getOrElse { println(it) }
